src/vmm/src/resources.rs

```diff
@@ -278,3 +278,296 @@ impl VmResources {
+        assert_eq!(vm_resources.vm_config, aux_vm_config);
+
+        // Invalid vcpu count
+        aux_vm_config.vcpu_count = Some(0);
+        assert!(vm_resources.set_vm_config(aux_vm_config.clone()).is_err());
+        aux_vm_config.vcpu_count = Some(33);
+        assert!(vm_resources.set_vm_config(aux_vm_config.clone()).is_err());
+        aux_vm_config.vcpu_count = Some(32);
+
+        // Invalid mem_size_mib
+        aux_vm_config.mem_size_mib = Some(0);
+        assert!(vm_resources.set_vm_config(aux_vm_config.clone()).is_err());
+    }
+
+    #[test]
+    fn test_boot_config() {
+        let vm_resources = default_vm_resources();
+        let expected_boot_cfg = vm_resources.boot_config.as_ref().unwrap();
+        let actual_boot_cfg = vm_resources.boot_source().unwrap();
+
+        assert_eq!(actual_boot_cfg, expected_boot_cfg);
+    }
+
+    #[test]
+    fn test_set_boot_source() {
+        let tmp_file = TempFile::new().unwrap();
+        let cmdline = "reboot=k panic=1 pci=off nomodules 8250.nr_uarts=0";
+        let expected_boot_cfg = BootSourceConfig {
+            kernel_image_path: String::from(tmp_file.as_path().to_str().unwrap()),
+            initrd_path: Some(String::from(tmp_file.as_path().to_str().unwrap())),
+            boot_args: Some(cmdline.to_string()),
+        };
+
+        let mut vm_resources = default_vm_resources();
+        let boot_cfg = vm_resources.boot_source().unwrap();
+        let tmp_ino = tmp_file.as_file().metadata().unwrap().st_ino();
+
+        assert_ne!(boot_cfg.cmdline.as_str(), cmdline);
+        assert_ne!(boot_cfg.kernel_file.metadata().unwrap().st_ino(), tmp_ino);
+        assert_ne!(
+            boot_cfg
+                .initrd_file
+                .as_ref()
+                .unwrap()
+                .metadata()
+                .unwrap()
+                .st_ino(),
+            tmp_ino
+        );
+
+        vm_resources.set_boot_source(expected_boot_cfg).unwrap();
+        let boot_cfg = vm_resources.boot_source().unwrap();
+        assert_eq!(boot_cfg.cmdline.as_str(), cmdline);
+        assert_eq!(boot_cfg.kernel_file.metadata().unwrap().st_ino(), tmp_ino);
+        assert_eq!(
+            boot_cfg
+                .initrd_file
+                .as_ref()
+                .unwrap()
+                .metadata()
+                .unwrap()
+                .st_ino(),
+            tmp_ino
+        );
+    }
+
+    #[test]
+    fn test_set_block_device() {
+        let mut vm_resources = default_vm_resources();
+        let mut new_block_device_cfg = vm_resources.block.config_list.get(0).unwrap().clone();
+        let tmp_file = TempFile::new().unwrap();
+        new_block_device_cfg.drive_id = "block3".to_string();
+        new_block_device_cfg.path_on_host = tmp_file.as_path().to_path_buf();
+        assert_eq!(vm_resources.block.config_list.len(), 1);
+        vm_resources.set_block_device(new_block_device_cfg).unwrap();
+        assert_eq!(vm_resources.block.config_list.len(), 2);
```

Comment:
For checking the `insert` functionality, it is a bit misleading here that you insert a new block device by changing the id of an existing one. Maybe a comment would be useful here or you could just create a new `BlockDeviceConfig` instance instead. Also, `block3` -> `block2`.

Reply:
The insertion of a new block device configuration happens after cloning and modifying the clone of the existing block device configuration, to be able to insert it. I find it less repetitive this way.

Reply:
I found it much more easier to follow with a simple insertion. I do not have a strong preference here, but please add a comment :D.